cli: add debug merge-logs command to merge log files into a single stream.

ajwerner opened the pull request:
Debugging distributed systems is challenging, often requiring correlating logs
from many servers. This PR add the new `cockroach debug merge-logs` command
which simplifies the workflow for combining logs from multiple servers. The new
command takes a list of log files or glob patterns which expand to log file
names and combines them into a single sorted stream. By default only logs from
programs named "cockroach" are considered though this is controllable from the
`program` flag which takes a regular expression over program name.

When logs are merged, a prefix of the form `{{ .Details.Host }}>` are added to
the front of log messages. This can be overridden by supplying a template string
to the `prefix` flag which is evaluated on a *log.FileInfo struct.

The command accepts unexpanded glob patterns to enable use in situations where
thousands of files exist and shell expansion would overflow MAX_ARG_STRLEN. The
command provides functionality to efficiently explore specific time ranges by
only reading files which might contain entries in the specified interval and by
using binary search to find an appropriate offset within files. It also provides
a mechanism for regular expression filtering of log line messages, something
which is difficult to accomplish with just grep.

Within a given set of files for the same node, log messages which occur out of
time order are "ratcheted" to be equal to the last time seen from that node.

More thought could be given to how this command behaves in situations where a
host time may have moved backwards appreciably such that file timestamps are not
properly ordered.

Proposed in favor of https://github.com/cockroachdb/cockroach/pull/32435.

Release note (cli change): Adds debug merge-logs command to combine logs from multiple nodes.

Commit:
cli: add debug merge-logs command to merge log files into a single stream.

Debugging distributed systems is challenging, often requiring correlating logs
from many servers. This PR add the new `cockroach debug merge-logs` command
which simplifies the workflow for combining logs from multiple servers. The new
command takes a list of log files or glob patterns which expand to log file
names and combines them into a single sorted stream. By default only logs from
programs named "cockroach" are considered though this is controllable from the
`program` flag which takes a regular expression over program name.

When logs are merged, a prefix of the form `{{ .Details.Host }}>` are added to
the front of log messages. This can be overridden by supplying a template string
to the `prefix` flag which is evaluated on a *log.FileInfo struct.

The command accepts unexpanded glob patterns to enable use in situations where
thousands of files exist and shell expansion would overflow MAX_ARG_STRLEN. The
command provides functionality to efficiently explore specific time ranges by
only reading files which might contain entries in the specified interval and by
using binary search to find an appropriate offset within files. It also provides
a mechanism for regular expression filtering of log line messages, something
which is difficult to accomplish with just grep.

Within a given set of files for the same node, log messages which occur out of
time order are "ratcheted" to be equal to the last time seen from that node.

More thought could be given to how this command behaves in situations where a
host time may have moved backwards appreciably such that file timestamps are not
properly ordered.

Proposed in favor of https://github.com/cockroachdb/cockroach/pull/32435.

Release note (cli change): Adds debug merge-logs command to combine logs from multiple nodes.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/32790)
<!-- Reviewable:end -->

tbg reviewed (commented):
Great, can't wait to use this! Will give a proper review later, but one question I already have is that not all of the test data seems to be there (or reviewable is tricking me somehow):

![✂-1](https://files.reviewableusercontent.io/images/reviewable/5076964/dRohVdxGM9OYXCLceiBQCbcI.png)

> Reviewed 7 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-:-LSp8zqD3eh4f9A_jRD5:b-65fwf4)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cli/debug_merge_logs_test.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LSp8QVr2QP1qPXvYcla:-LSp8QVr2QP1qPXvYclb:b-788aey) ([raw file](https://github.com/cockroachdb/cockroach/blob/beaa5df61f560132da9536ab63fb7d77760458bc/pkg/cli/debug_merge_logs_test.go#L30)):*
> ```Go
> 	flags    []string
> 	args     []string
> 	expected string
> ```

I'd also put the expected output into files.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Thanks! I think they were being gitignored, I was perplexed why it was passing locally and not in teamcity. I'll move the outputs to files now. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-:-LSpBH0Z39O8E9ycc4GM:b-ttecqf)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

ajwerner commented:
I also found the merged logs to be difficult to read so I created a separate tool https://github.com/ajwerner/logcolor/ which accepts the merged stream and colorizes it based on the prefix by hashing to a color.

![screenshot from 2018-12-03 12-53-39](https://user-images.githubusercontent.com/1839234/49391913-fbb4ff00-f6fa-11e8-9025-d4aa99b37cf5.png)

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Generally looks like what I expected, though I didn't go over this with a fine-toothed comb. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-:-LSpFP_yB3_iPlOQU7xr:b-p7lynm)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/cli/debug_merge_logs.go, line 44 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LSpDAiU3RWUYf5n-_PC:-LSpDAiU3RWUYf5n-_PD:bkvy9t6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3478ee513f1c00b5052aa1b3ccaddda04f3b45a/pkg/cli/debug_merge_logs.go#L44)):*
> ```Go
> 
> func newMergedStream(files []fileInfo, from, to time.Time) *mergedStream {
> 	// TODO: think about clock movement and PID
> ```

Nit: we put author names in parens after the TODO. `TODO(ajwerner)`.
___
*[pkg/cli/debug_merge_logs.go, line 53 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LSpF8ao0Nk2w2zoi2mO:-LSpF8ao0Nk2w2zoi2mP:b-4x9gcj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3478ee513f1c00b5052aa1b3ccaddda04f3b45a/pkg/cli/debug_merge_logs.go#L53)):*
> ```Go
> 	// TODO(ajwerner): could do the peek in parallel though most of the cost
> 	// has been mitigated with the binary search.
> 	addLogStream := func() {
> ```

There is a bit of complexity in this code due to this concatenated per-node log stream. Did you consider just putting all of the logs in the heap? The downside to that is that you wouldn't want to keep all of the logs open, but I think you could avoid that problem by peeking at the first record in each log first.
___
*[pkg/cli/debug_merge_logs.go, line 72 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LSpDRJG0XsWtGuGwzia:-LSpDRJG0XsWtGuGwzib:b-lvpgm2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3478ee513f1c00b5052aa1b3ccaddda04f3b45a/pkg/cli/debug_merge_logs.go#L72)):*
> ```Go
> 		}
> 		// Overwrite the previous file if there is one and the current file starts
> 		// before from.
> ```

I've noticed instances of a single host producing logs with overlapping timestamps. Is that what you're trying to handle here? It will be very unfortunate if this code accidentally drops a log.
___
*[pkg/cli/debug_merge_logs_test.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LSp8QVr2QP1qPXvYcla:-LSpD-6Q2JvUsYD2Naqn:b-34b270) ([raw file](https://github.com/cockroachdb/cockroach/blob/beaa5df61f560132da9536ab63fb7d77760458bc/pkg/cli/debug_merge_logs_test.go#L30)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I'd also put the expected output into files.
</blockquote></details>

Agreed. If you want to get fancy you can use `testutils.datadriven`, though that might be overkill here.
___
*[pkg/util/log/file.go, line 124 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LSpCm_8A-vGq3k6JeAz:-LSpCm_9-9Z74XZqTa7M:b8msu6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3478ee513f1c00b5052aa1b3ccaddda04f3b45a/pkg/util/log/file.go#L124)):*
> ```Go
> 
> // NewFileInfo constructs a FileInfo from FileDetails and os.FileInfo.
> func NewFileInfo(details FileDetails, info os.FileInfo) FileInfo {
> ```

Nit: We usually reserve `New` for when a function returns a pointer and use the term `Make` when it returns a value. So this function should be named `MakeFileInfo`.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32790)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cli/debug_merge_logs.go, line 53 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LSpF8ao0Nk2w2zoi2mO:-LSpIiopA6sOfof4gtKO:b-njf5b1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3478ee513f1c00b5052aa1b3ccaddda04f3b45a/pkg/cli/debug_merge_logs.go#L53)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

There is a bit of complexity in this code due to this concatenated per-node log stream. Did you consider just putting all of the logs in the heap? The downside to that is that you wouldn't want to keep all of the logs open, but I think you could avoid that problem by peeking at the first record in each log first.
</blockquote></details>

I mostly didn't want to put all the files into a single heap so I didn't have to open them all then need a semaphore or something. Adding the logic to peek at the first record, then close the file and then reopen seems like a similar amount of complexity. I also figured that it might be possible to reorder log files for a node such that all files with the same PID are contiguous or something in the case of clock changes but given I didn't type that code and I'm not sure how it would actually work, it seems like a case of YAGNI. 

The downside of the current approach is if there are two files with the same timestamp then it's possible we'll look at them out of order, which seems like a big enough downside to change this logic. Will revisit. 
___
*[pkg/cli/debug_merge_logs.go, line 72 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LSpDRJG0XsWtGuGwzia:-LSpJ-yeB4QEMtfkpxBW:b-f0g1gp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3478ee513f1c00b5052aa1b3ccaddda04f3b45a/pkg/cli/debug_merge_logs.go#L72)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I've noticed instances of a single host producing logs with overlapping timestamps. Is that what you're trying to handle here? It will be very unfortunate if this code accidentally drops a log.
</blockquote></details>

Can you help me to better understand when this situation occurs? If it does happen regularly then it would warrant opening all of the files and peeking at them here. 
___
*[pkg/cli/debug_merge_logs_test.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LSp8QVr2QP1qPXvYcla:-LSpJiFZEme09KHb3r3a:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/beaa5df61f560132da9536ab63fb7d77760458bc/pkg/cli/debug_merge_logs_test.go#L30)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Agreed. If you want to get fancy you can use `testutils.datadriven`, though that might be overkill here.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32790)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cli/debug_merge_logs.go, line 53 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LSpF8ao0Nk2w2zoi2mO:-LSpQphw9jvedQjaZoDN:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3478ee513f1c00b5052aa1b3ccaddda04f3b45a/pkg/cli/debug_merge_logs.go#L53)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I mostly didn't want to put all the files into a single heap so I didn't have to open them all then need a semaphore or something. Adding the logic to peek at the first record, then close the file and then reopen seems like a similar amount of complexity. I also figured that it might be possible to reorder log files for a node such that all files with the same PID are contiguous or something in the case of clock changes but given I didn't type that code and I'm not sure how it would actually work, it seems like a case of YAGNI. 

The downside of the current approach is if there are two files with the same timestamp then it's possible we'll look at them out of order, which seems like a big enough downside to change this logic. Will revisit. 
</blockquote></details>

Ack.
___
*[pkg/cli/debug_merge_logs.go, line 72 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LSpDRJG0XsWtGuGwzia:-LSpQebz4PwVn_HaV9Xb:b-jzb3eh) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3478ee513f1c00b5052aa1b3ccaddda04f3b45a/pkg/cli/debug_merge_logs.go#L72)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Can you help me to better understand when this situation occurs? If it does happen regularly then it would warrant opening all of the files and peeking at them here. 
</blockquote></details>

I have no idea when it occurs or why it occurs. It is just something I've observed when running various roachtests and then looking at the logs. I think most recently I saw it with the `disk-full` roachtest.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32790)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cli/debug_merge_logs.go, line 53 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LSpF8ao0Nk2w2zoi2mO:-LSqI0vw2Wl2nYdkMPeQ:budo0n4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3478ee513f1c00b5052aa1b3ccaddda04f3b45a/pkg/cli/debug_merge_logs.go#L53)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ack.
</blockquote></details>

I was wrong about the filename ts resolution. It's definitely on a per second basis. The number after the dot is the PID. I'll change the logic to peek at log entries per file. 


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32790)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cli/debug_merge_logs.go, line 53 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LSpF8ao0Nk2w2zoi2mO:-LStVE_VEFu_mAq2UybA:bgqds2u) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3478ee513f1c00b5052aa1b3ccaddda04f3b45a/pkg/cli/debug_merge_logs.go#L53)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I was wrong about the filename ts resolution. It's definitely on a per second basis. The number after the dot is the PID. I'll change the logic to peek at log entries per file. 
</blockquote></details>

Oh, that probably explains what I was seeing. If I recall correctly, the timestamp was identical, but the portion after the decimal point was lower in the second log file. This is in a test that was rapidly restarting a node so that makes sense.


<!-- Sent from Reviewable.io -->

Commit:
cli: clean up and add buffering to debug merge-logs

This commit moves the logic of writing and reading of files into separate
goroutines. The change seems to increase the throughput of debug merge-logs by
almost 4x from ~16MB/s to ~60MB/s on my local laptop when merging 4GB of logs
from 102 hosts into /dev/null.

Release note: None

ajwerner reviewed (commented):
I got a little carried away tuning the performance for this but I got a little frustrated at how long it was taking to scan and filter large amounts of data. It feels pretty fast now.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-:-LSzmA1ZBOiSrcPY690O:bjou3cy)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cli/debug_merge_logs.go, line 53 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LSpF8ao0Nk2w2zoi2mO:-LSzltOP56IbvUQ78602:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3478ee513f1c00b5052aa1b3ccaddda04f3b45a/pkg/cli/debug_merge_logs.go#L53)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Oh, that probably explains what I was seeing. If I recall correctly, the timestamp was identical, but the portion after the decimal point was lower in the second log file. This is in a test that was rapidly restarting a node so that makes sense.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

The implementation looks nice and clean. You mention performance tuning. Is the performance good with that channel based pipeline?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-:-LT0AgI3DsSwOBs2avvW:b-ywo8i9)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/cli/debug.go, line 1291 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LT08quX2en4m7BTW3Ev:-LT08quYEb8JyFTQWAkV:b-yyds4v) ([raw file](https://github.com/cockroachdb/cockroach/blob/3015781f2a9bbeac23deb83305f91830b98f387f/pkg/cli/debug.go#L1291)):*
> ```Go
> 		return time.Time{}, nil
> 	}
> 	t, err := time.Parse("060102 15:04:05.999999", ts)
> ```

I wonder if this should support alternative time formats, as this one is pretty verbose. Feel free to leave this suggestion to a future PR.
___
*[pkg/cli/debug_merge_logs.go, line 73 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LT09Of00qT38KUuBEzL:-LT09Of00qT38KUuBEzM:b81wdqy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3015781f2a9bbeac23deb83305f91830b98f387f/pkg/cli/debug_merge_logs.go#L73)):*
> ```Go
> 	entryChan := make(chan *entryInfo, chanSize)  // read -> process
> 	msgChan := make(chan *bytes.Buffer, chanSize) // process -> bufferWrites
> 	writeChan := make(chan *bytes.Buffer)         // bufferWrites -> write
> ```

Whoa, channels! I haven't fully grok'd this yet, but I suspect this might be an overuse of channels. Channels are a neat way to structure a pipeline like this, but they can actually be a bottleneck if the processing is minimal as I suspect it is here. Or perhaps it isn't. On the plus side, this pipeline is nice and clean.

Looking at this a bit more, as long as the performance is good, I like the structure of this code.
___
*[pkg/cli/debug_merge_logs.go, line 79 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LT0B-T4AQ2KRFBM8wtW:-LT0B-T4AQ2KRFBM8wtX:bzic3co) ([raw file](https://github.com/cockroachdb/cockroach/blob/3015781f2a9bbeac23deb83305f91830b98f387f/pkg/cli/debug_merge_logs.go#L79)):*
> ```Go
> 		for e, ok := s.peek(); ok; e, ok = s.peek() {
> 			select {
> 			case entryChan <- &entryInfo{Entry: e, FileInfo: s.fileInfo()}:
> ```

Does `entryInfo` need to be a pointer here? I think you can use `chan entryInfo` and avoid the allocation. Probably doesn't matter and this is just a knee-jerk comment on my part.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
It's now able to pin all the CPUs on my laptop to 100%. The biggest wins were copying the regexp in log.NewEntryDecoder and caching the template rendered prefix. Now the vast majority of time is spend decoding log entries, as it should be.  I think I'm done tuning this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-:-LT0Oc1i4mJAYV37dcJt:b-nv6j30)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cli/debug.go, line 1291 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LT08quX2en4m7BTW3Ev:-LT3CPwt740Eks5w0o8E:b-udfsmw) ([raw file](https://github.com/cockroachdb/cockroach/blob/3015781f2a9bbeac23deb83305f91830b98f387f/pkg/cli/debug.go#L1291)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I wonder if this should support alternative time formats, as this one is pretty verbose. Feel free to leave this suggestion to a future PR.
</blockquote></details>

I had some other time stamp formats before that were relative to now. I'll add them back in a different PR.
___
*[pkg/cli/debug_merge_logs.go, line 73 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LT09Of00qT38KUuBEzL:-LT0MXO10u19TFAoNz0F:b-jop110) ([raw file](https://github.com/cockroachdb/cockroach/blob/3015781f2a9bbeac23deb83305f91830b98f387f/pkg/cli/debug_merge_logs.go#L73)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Whoa, channels! I haven't fully grok'd this yet, but I suspect this might be an overuse of channels. Channels are a neat way to structure a pipeline like this, but they can actually be a bottleneck if the processing is minimal as I suspect it is here. Or perhaps it isn't. On the plus side, this pipeline is nice and clean.

Looking at this a bit more, as long as the performance is good, I like the structure of this code.
</blockquote></details>

This pipeline lets means that all of the blocking IO happens in its own goroutine. My thinking upon reading this was that the goroutine that probably is wasteful here is the one that does "processing" which was actually the first one that existed. Initially I had just the reading goroutine and the processing one which then wrote directly to out. The problem with that is to make it performant I wrapped out in a bufio.Writer which had the problem of sometimes writing just portions of messages to stdout, which is less than pleasant. Anyway, I went and dropped one of the levels of goroutines and the whole buffer copy thing and it lead to slightly higher throughput.
___
*[pkg/cli/debug_merge_logs.go, line 79 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LT0B-T4AQ2KRFBM8wtW:-LT0BsXxAQj7mXOpXDLO:b-gwturl) ([raw file](https://github.com/cockroachdb/cockroach/blob/3015781f2a9bbeac23deb83305f91830b98f387f/pkg/cli/debug_merge_logs.go#L79)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Does `entryInfo` need to be a pointer here? I think you can use `chan entryInfo` and avoid the allocation. Probably doesn't matter and this is just a knee-jerk comment on my part.
</blockquote></details>

It doesn't seem to make a difference. Anecdotally I swapped it because it seemed to have higher throughput as a pointer with my theory being that the channel write is cheaper but running it a few more times each way it seems like noise. 


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
:shipit: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-:-LT3_ymz9s9lt7HFmXpw:bizlbj6)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cli/debug.go, line 1291 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LT08quX2en4m7BTW3Ev:-LT3_w6C9ycSkfM3-aRc:bb73djm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3015781f2a9bbeac23deb83305f91830b98f387f/pkg/cli/debug.go#L1291)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I had some other time stamp formats before that were relative to now. I'll add them back in a different PR.
</blockquote></details>

SGTM


<!-- Sent from Reviewable.io -->

tbg commented:
Haven't gotten around to reviewing, but I've used it yesterday and had a
blast. One comment I had is that some users have different formats for
grabbing the logs (they log to stderr and that feeds into their logging
subsystem) in which case you'll get files, but they may be all over the
place.

Here's a particular example:

x/var/log/cockroachdb/@400000005bf6647d0d2d8324.s
x/var/log/cockroachdb/@400000005bfac77e231b1d2c.s
x/var/log/cockroachdb/@400000005c002e9f0873db44.s
x/var/log/cockroachdb/@400000005c060a9507d98774.s
x/var/log/cockroachdb/current

`current` is the current log file and the other ones are lz compressed
older ones. Now I definitely don't expect `merge-logs` to realize the
compression and to handle it transparently, but I should only be a `for f
in *.s; do $magic_decompress $f; done` away from being able to use
`merge-logs` with the right glob. Is that true in the code today as written
or do we rely on the log files having been written by CRDB and not renamed?
Doesn't need to block this PR and I'm happy to help out if you want this
off your plate (I also have logs in this format that I can point you at
privately).

ajwerner reviewed (commented):
There's currently not a good story about dealing with different log file names. I am happy to extend the functionality to support different naming schemes. I created https://github.com/cockroachdb/cockroach/issues/32919 to track, please add commentary there. 

Somewhat unrelatedly, before I started this I was thinking that it might be nice to allow users to merge logs inside of a compressed archive without first needing to do decompression. I could imagine supporting input patterns that maybe look like `cockroach debug merge-logs whole-cluster-logs.zip:logs/*/*` where the tool detects the archive and expands the glob inside of it. The reason I  got discouraged from jumping on this straight away was the tight coupling of glob expansion logic in filepath.Glob to the filesystem APIs. This functionality too should come in another PR if at all, just proposing it to see if there's interest. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-:-LT4ZOjyD2GDTOHW7rY5:beykxqg)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

tbg commented on pkg/cli/debug_merge_logs.go:
```diff
@@ -0,0 +1,556 @@
+	if err := g.Wait(); err != nil {
+		return nil, err
+	}
+	filtered := res[:0] // Filter the nil streams.
```
inconsequential nit, but thought you'd generally like to know: Go comment notation uses proper punctuation when the comment is on its own line, but not in inline comments, so this would technically be `// filter nil streams`. I don't know that I love this rule, though. Happy to ignore it.

ajwerner reviewed (commented):
TFYR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-:-LT8enRSA0H8m8mMfVcL:b-814cj3)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cli/debug_merge_logs.go, line 190 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/32790#-LSouaa31srn2jRVch0M-r5-190:-LT8bPR03NtSY8ZgMRJA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f5fb575bf02e6d1107704ef9da45d778d3d2d81/pkg/cli/debug_merge_logs.go#L190)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

inconsequential nit, but thought you'd generally like to know: Go comment notation uses proper punctuation when the comment is on its own line, but not in inline comments, so this would technically be `// filter nil streams`. I don't know that I love this rule, though. Happy to ignore it.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner commented:
bors r+
